lib/stdlib/src/supervisor.erl

```diff
@@ -1516,11 +1516,38 @@ shutdown(#child{pid=Pid, shutdown=Time} = Child) ->
                     normal when not (?is_permanent(Child)) ->
                         ok;
                     Reason ->
+                        %% killed is expected to become an error as
+                        %% this function clause tries to terminate
+                        %% process not brutally
                         {error, Reason}
                 end
         end
     end.
 
+unlink_flush(Pid, NotExitReason) when NotExitReason == noproc ->
+                                % ? check only for noproc?
+                                % noconnection should remain an error
+                                %% NotExitReason == noconnection ->
+    %% If race between orderly shutdown from supervisor termination
+    %% and worker termination occurs make sure to report
+    %% correct error reason if possible
+    {links, Links} = process_info(self(), links),
+    case lists:member(Pid, Links) of
+        true ->
+            receive
+                {'EXIT', Pid, Reason} ->
+                    Reason
+            end;
+        false ->
+            %% naughty child - unlinked itself from supervisor
+            %% we need after section to handle such child
+            receive
+                {'EXIT', Pid, Reason} ->
+                    Reason
+            after 0 ->
+                    NotExitReason
+            end
+    end;
 unlink_flush(Pid, DefaultReason) ->
```

Comment:
I think we need to do something like this so as we want to change noproc = Reaosn to be a normal case as that is the race.

```
unlink_flush(Pid, Reaon) ->
    %% We call unlink in order to guarantee that the 'EXIT' has arrived
    %% from the dead process. See the unlink docs for details.
    {links, Links} = process_info(self(), links),
    case lists:member(Pid, Links) of
        true ->
            do_unlink_flush(Pid, Reason);
        false  -> %% Naughty child
            flush(Pid, Reason)
    end.

do_unlink_flush(Pid, noproc = Reason) ->
    unlink(Pid),
    receive
        {'EXIT', Pid, Reason} ->
            normal
    end;
do_unlink_flush(Pid, Reason) ->
    unlink(Pid),
    receive
        {'EXIT', Pid, Reason} ->
            Reason
    end.

flush(Pid, NoExit) ->
    receive
        {'EXIT', Pid, Reason} ->
            Reason
    after 0 ->
            NoExit
    end.
```